### PR TITLE
Kirill vecdb stuff

### PR DESCRIPTION
**Pull request by olegklimov**

### CODE REVIEW

File: src/integrations/integr_postgres.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+impl ToolPostgres {
+    pub fn new_if_configured(integrations_value: &serde_yaml::Value) -> Option<Self> {
+        let postgres = integrations_value.get("postgres")?;
+        let connection_string = postgres.get("connection_string")?.as_str()?.to_string();
+        let psql_binary_path = postgres.get("psql_binary_path")?.as_str()?;
+
+        Some(ToolPostgres {
+            connection_string,
+            psql_binary_path: PathBuf::from(psql_binary_path),
+        })
```
Comment (olegklimov) [inline src/integrations/integr_postgres.rs:29]: Deserializing ToolPostgres will produce better errors

File: src/integrations/integr_postgres.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    async fn run_psql_command(&self, query: &str) -> Result<String, String> {
+        let mut cmd = Command::new(&self.psql_binary_path);
+        cmd.arg(&self.connection_string)
+            .arg("ON_ERROR_STOP=1")
+            .arg("-c")
+            .arg(query);
+
+        let output = cmd.output().await
+            .map_err(|e| format!("Failed to execute psql command: {}", e))?;
```
Comment (olegklimov) [inline src/integrations/integr_postgres.rs:40]: No timeout? We need timeout.

File: src/tools/tools_description.rs [hunk 3]
```diff
@@ -274,6 +278,17 @@ tools:
...
+  - name: "postgres"
+    agentic: true
+    experimental: true
+    description: "Execute PostgreSQL queries using psql command-line tool."
+    parameters:
+      - name: "command"
+        type: "string"
+        description: "Examples: 'SELECT * FROM table_name', 'SELECT * FROM table_name WHERE column_name = value'" 
+    parameters_required:
+      - "columns"
```
Comment (olegklimov) [inline src/tools/tools_description.rs:291]: columns

File: src/tools/tools_description.rs [hunk 4]
```diff
@@ -274,6 +278,17 @@ tools:
         description: "Examples: 'python -m pdb script.py', 'break module_name.function_name', 'break 10', 'continue', 'print(variable_name)', 'list', 'quit'"
     parameters_required:
       - "command"
+      
+  - name: "postgres"
+    agentic: true
+    experimental: true
+    description: "Execute PostgreSQL queries using psql command-line tool."
+    parameters:
+      - name: "command"
```
Comment (olegklimov) [inline src/tools/tools_description.rs:287]: query

File: src/integrations/mod.rs [hunk 5]
```diff
@@ -26,4 +28,9 @@ commands_deny:
 #pdb:
 #  python_path: "/opt/homebrew/bin/python3"  # Uncomment to set a custom python path, defaults to "python3"
 
+# --- Postgres integration ---
+#postgres:
+#  psql_binary_path: "/path/to/psql"  # Uncomment to set a custom path for the psql binary, defaults to "psql"
```
Comment (olegklimov) [inline src/integrations/mod.rs:33]: There is no default really

File: src/integrations/integr_postgres.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        let mut results = vec![];
+        results.push(ContextEnum::ChatMessage(ChatMessage {
+            role: "tool".to_string(),
+            content: ChatContent::SimpleText(serde_json::to_string(&result).unwrap()),
+            tool_calls: None,
+            tool_call_id: tool_call_id.clone(),
+            ..Default::default()
+        }));
+        Ok((true, results))
```
Comment (olegklimov) [inline src/integrations/integr_postgres.rs:74]: can psql write output in json?

File: src/tools/tools_execute.rs [hunk 7]
```diff
@@ -132,6 +132,11 @@ pub async fn run_tools(
                     generated_tool.push(tool_failed_message);
                     continue;
                 }
+                let (needs_confirmation, confirmation_reason) = command_should_be_confirmed_by_user(&command_to_match, &rules.commands_need_confirmation);
+                if needs_confirmation {
+                    generated_tool.push(tool_answer(confirmation_reason, t_call.id.to_string()));
+                    continue;
+                }
```
Comment (olegklimov) [inline src/tools/tools_execute.rs:139]: shouldn't be there

File: src/vecdb/vdb_highlev.rs [hunk 8]
```diff
@@ -296,14 +291,12 @@ pub async fn memories_add(
...
-) -> Result<(), String> {
-    let vectorizer_service = {
-        let vec_db_guard = vec_db.lock().await;
-        let vec_db = vec_db_guard.as_ref().ok_or("VecDb is not initialized")?;
-        vec_db.vectorizer_service.clone()
-    };
+
+pub async fn memories_block_until_vectorized_from_vectorizer(
+    vectorizer_service: Arc<AMutex<FileVectorizerService>>, 
+    max_blocking_time_ms: usize) -> Result<(), String>  {
```
Comment (olegklimov) [inline src/vecdb/vdb_highlev.rs:297]: unreadable code because of { at the end

File: src/integrations/integr_postgres.rs (hunk 6 above)
Comment (reymondzzzz) [reply src/integrations/integr_postgres.rs:74]: Nope 
image

File: src/vecdb/vdb_thread.rs [hunk 9]
```diff
@@ -195,31 +149,71 @@ async fn from_splits_to_vecdb_records_applying_cache(
...
+        let current_time = SystemTime::now();
+        let todo_len = {
+            let vservice_locked = vservice.lock().await;
+            let mut vecdb_todo_locked = vservice_locked.vecdb_todo.lock().await;
+            if let Some(msg) = vecdb_todo_locked.pop_front() {
+                match msg {
+                    MessageToVecdbThread::RegularDocument(doc) => {
+                        last_updated.insert(doc, current_time);
+                    }
+                    MessageToVecdbThread::ImmediatelyRegularDocument(_) |
```
Comment (olegklimov) [inline src/vecdb/vdb_thread.rs:192]: Ignores messages

File: src/vecdb/vdb_thread.rs [hunk 10]
```diff
@@ -195,31 +149,71 @@ async fn from_splits_to_vecdb_records_applying_cache(
...
     loop {
-        let (msg_to_me, files_unprocessed, vstatus_changed) = {
-            let mut qlocked = vecdb_immediate_q.lock().await;
-            let q_len = qlocked.len();
-            // CAREFUL: two locks, qlocked -> vstatus_locked
+        let mut msg_to_me: Option<MessageToVecdbThread> = None;
+        let current_time = SystemTime::now();
+        let todo_len = {
+            let vservice_locked = vservice.lock().await;
+            let mut vecdb_todo_locked = vservice_locked.vecdb_todo.lock().await;
```
Comment (olegklimov) [inline src/vecdb/vdb_thread.rs:186]: Two locks, looks like it could be one

File: src/vecdb/vdb_thread.rs [hunk 11]
```diff
@@ -301,37 +295,39 @@ async fn vectorize_thread(
...
-                        let done = vstatus_locked.state == "done";
-                        if !done {
-                            vstatus_locked.files_unprocessed = 0;
-                            vstatus_locked.files_total = 0;
-                            vstatus_locked.state = "done".to_string();
-                            info!(
-                                "vectorizer since start {} API calls, {} vectors",
-                                vstatus_locked.requests_made_since_start, vstatus_locked.vectors_made_since_start
-                            );
+                    if last_updated.is_empty() {
```
Comment (olegklimov) [inline src/vecdb/vdb_thread.rs:298]: you can use
match() {
  _ if ... {
to shift code back to the left

File: src/vecdb/vdb_thread.rs [hunk 12]
```diff
@@ -1,88 +1,42 @@
...
-use crate::files_in_workspace::{Document, is_path_to_enqueue_valid};
+use crate::files_in_workspace::{is_path_to_enqueue_valid, Document};
 use crate::global_context::GlobalContext;
-use crate::knowledge::{MemoriesDatabase, vectorize_dirty_memories};
-use crate::vecdb::vdb_lance::VecDBHandler;
-use crate::vecdb::vdb_structs::{VecdbRecord, SplitResult, VecdbConstants, VecDbStatus, SimpleTextHashVector};
+use crate::knowledge::{vectorize_dirty_memories, MemoriesDatabase};
 use crate::vecdb::vdb_cache::VecDBCache;
+use crate::vecdb::vdb_lance::VecDBHandler;
+use crate::vecdb::vdb_structs::{SimpleTextHashVector, SplitResult, VecDbStatus, VecdbConstants, VecdbRecord};
```
Comment (olegklimov) [inline src/vecdb/vdb_thread.rs:19]: Only order of imports changes, useless change

File: src/cached_tokenizers.rs [hunk 13]
```diff
@@ -150,6 +150,7 @@ pub async fn cached_tokenizer(
     try_download_tokenizer_file_and_open(&client2, http_path.as_str(), api_key.clone(), &to).await?;
     info!("loading tokenizer \"{}\"", to.display());
     let mut tokenizer = Tokenizer::from_file(to).map_err(|e| format!("failed to load tokenizer: {}", e))?;
+    let _ = tokenizer.with_truncation(None);
```
Comment (olegklimov) [inline src/cached_tokenizers.rs:153]: "The with_truncation function in the context of tokenizers is used to enable or disable truncation of the tokenized sequence. When truncation is enabled, sequences longer than a specified maximum length will be truncated to fit within that limit. The None parameter likely indicates that there is no specific truncation configuration provided, meaning default settings might be used."
what does it really mean?

File: src/vecdb/vdb_thread.rs [hunk 14]
```diff
@@ -362,7 +362,7 @@ async fn vectorize_thread(
         }
 
         let file_splitter = AstBasedFileSplitter::new(constants.splitter_window_size);
-        let mut splits = file_splitter.vectorization_split(&doc, None, gcx.clone(), constants.vectorizer_n_ctx).await.unwrap_or_else(|err| {
+        let mut splits = file_splitter.vectorization_split(&doc, constants.tokenizer.clone(), gcx.clone(), constants.vectorizer_n_ctx).await.unwrap_or_else(|err| {
```
Comment (olegklimov) [inline src/vecdb/vdb_thread.rs:365]: are making this 10 times slower again?

File: src/ast/chunk_utils.rs [hunk 15]
```diff
@@ -15,6 +15,49 @@ pub fn official_text_hashing_function(s: &String) -> String {
...
+    if let Some(ids) = ids {
+        if ids.len() <= tokens_limit {
+            return vec![line.clone()];
+        }
+        
+        let chunks: Vec<Vec<u32>> = ids.chunks(tokens_limit).map(|chunk| Vec::from(chunk)).collect();
+        let mut result: Vec<String> = Vec::new();
+        {
+            let tokenizer_locked = &tokenizer.unwrap();
+            let tokenizer_locked = tokenizer_locked.read().unwrap();
```
Comment (olegklimov) [inline src/ast/chunk_utils.rs:41]: ask chat to simplify this, looks like this can be shorter

**Pull request closed without merging**